lib/ansible/plugins/inventory/netbox.py

```diff
@@ -0,0 +1,365 @@
+            default: True
+            type: boolean
+        token:
+            required: True
+            description: NetBox token.
+            env:
+                # in order of precedence
+                - name: NETBOX_TOKEN
+                - name: NETBOX_API_KEY
+        group_by:
+            description: Keys used to create groups.
+            type: list
+            choices:
+                - sites
+                - tenants
+                - racks
+                - device_roles
+                - device_types
+                - manufacturers
+            default: []
+        query_filters:
+            description: List of parameters passed to the query string (Multiple values may be separated by commas)
+            type: list
+        timeout:
+            description: Timeout for Netbox requests in seconds
+            type: int
+            default: 60
+'''
+
+EXAMPLES = '''
+# netbox_inventory.yml file in YAML format
+# Example command line: ansible-inventory --list -i netbox_inventory.yml
+
+plugin: netbox
+api_endpoint: http://localhost:8000
+group_by:
+  - device_roles
+query_filters:
+  - role: network-edge-router
+'''
+
+import json
+from sys import version as python_version
+
+from ansible.plugins.inventory import BaseInventoryPlugin
+from ansible.module_utils.ansible_release import __version__ as ansible_version
+from ansible.errors import AnsibleError
+from ansible.module_utils._text import to_text
+from ansible.module_utils.urls import open_url
+from ansible.module_utils.six.moves.urllib.parse import urljoin, urlencode
+from ansible.module_utils.compat.ipaddress import ip_interface
+try:
+    from __main__ import display
+except ImportError:
+    from ansible.utils.display import Display
+    display = Display()
+
+ALLOWED_DEVICE_QUERY_PARAMETERS = (
+    "asset_tag",
+    "cluster_id",
+    "device_type_id",
+    "has_primary_ip",
+    "is_console_server",
+    "is_full_depth",
+    "is_network_device",
+    "is_pdu",
+    "mac_address",
+    "manufacturer",
+    "manufacturer_id",
+    "model",
+    "name",
+    "platform",
+    "platform_id",
+    "position",
+    "rack_group_id",
+    "rack_id",
+    "role",
+    "role_id",
+    "serial",
+    "site",
+    "site_id",
+    "status",
+    "tag",
+    "tenant",
+    "tenant",
+    "tenant_id",
+    "virtual_chassis_id",
+)
+
+
+class InventoryModule(BaseInventoryPlugin):
+    NAME = 'netbox'
+
+    def _fetch_information(self, url):
+
+        response = open_url(url, headers=self.headers, timeout=self.timeout)
+
+        try:
+            raw_data = to_text(response.read(), errors='surrogate_or_strict')
+        except UnicodeError:
+            raise AnsibleError("Incorrect encoding of fetched payload from Online servers")
+
+        try:
+            return json.loads(raw_data)
+        except ValueError:
+            raise AnsibleError("Incorrect JSON payload")
```

Comment:
I would propose to give the json paylod in the error message, as it would help for debugging